### PR TITLE
Wayland: Make panel shutdown button work under wayland

### PR DESCRIPTION
**Pull request by lukefromdc**
*This uses systemd to be able to shutdown when root is logged in anywhere, possibly including display managers *If user login is only login shutdown and reboot work w/o systemd *Suspend requires systemd at this point
*Logind support would be welcome

### CODE REVIEW

**Comment by lukefromdc** [comment]
As usual, this is a wayland-only change and should have no effect on x11 behavior.

The wayland shutdown button works by trying `system("shutdown now") ` first for shutdown and ` system ("reboot now")` first for reboot, then systemd's versions of those commands if that fails. We do not have "suspend" as a low level system command, so we are forced to rely on systemd (or logind if anyone will write support for it) for suspend.

A lot of the action button functions (logout, force-quit) are not easy to implement in wayland as using system, systemd, or logind commands are not enough. 

Note that logging in/out under wayland works differently than in x11 but there are a number of standalone logout managers available for wayland. A panel launcher for one could perhaps be offered for one with very little effort if it is installed

**Comment by lukefromdc** [comment]
Request for more Travis credits submitted

**Comment by raveit65** [comment]
Suspend-disk or hibernate button is missing. Command is `systemctl hibernate`.
I have that configured on my system and i can test it.

**Comment by lukefromdc** [comment]
I can add that button easily enough. It should be either not added or greyed out 
when hibernation is not available, but to do that I need to find a way to detect that 
in the absence of a session manager. If that turns into a no-go, we could simply
have a dialog message come up when an unsupported button is clicked.

**Comment by raveit65** [comment]
On systemd boxes shutdown, reboot and halt commands are implemented by systemd, see manpages.
Are they `SysV` systems nowadays which use the old start system?
With `logind` support you mean systemd-logind or the elogind project? https://github.com/elogind/elogind
Systemd -logind provides the `loginctl` command. Not tested with wayland session but maybe `loginctl lock-session [ID...]` or `loginctl lock-sessions` can be used for lock the session?
See man loginctl for options.

**Comment by raveit65** [comment]
> I can add that button easily enough. It should be either not added or greyed out
when hibernation is not available, but to do that I need to find a way to detect that
in the absence of a session manager.

With X11 and systemd /etc/systemd/sleep.conf is used.
Without  `AllowHibernation=yes` the button isn't visible if i recall correctly.
I use this config:
```
[Sleep]
AllowSuspend=yes
AllowHibernation=yes
AllowSuspendThenHibernate=yes
AllowHybridSleep=yes
#SuspendMode=
#SuspendState=mem standby freeze
HibernateMode=platform shutdown
```
from this guide https://www.ctrl.blog/entry/fedora-hibernate.html

**Comment by lukefromdc** [comment]
The existing x11 code makes some dbus calls to connect to systems, which I am not familiar with writing new code for yet.

**Comment by lukefromdc** [comment]
Tells me out to set up hibernation (this is an encrypted machine with 40G RAM so not practical) but not how to determine via systemd if it is enabled. The system() call lets me execute a command, but I would need to be able to tell from inside the C binary whether cat /sys/power/disk | grep "hibernate" returns a zero or non-zero length string at the least.

**Comment by lukefromdc** [comment]
I have the button working but on my machine w/o hiberation support clicking it does nothing

**Comment by raveit65** [comment]
Forget about loginctl. This doeasn't work with wayfire.

**Comment by lukefromdc** [comment]
Just forced-pushed the code as it sits now, the hibernate button is shown unconditionally.

**Comment by raveit65** [comment]
> I have the button working but on my machine w/o hiberation support clicking it does nothing

Sure you need a swap disk
the resume modul nedds to be added as dracut modul to dracut config
initramfs needs to be rebuild
A grub entry `GRUB_CMDLINE_LINUX="[…] resume=UUID=your-swap-identifier"`
If encrypted `GRUB_CMDLINE_LINUX="[…] rd.luks.uuid=your-swap-identifier"`
and an upddated /etc/systemd/sleep.conf
See my link.

But i am sure they are ToDo's for debian :)

**Comment by lukefromdc** [comment]
I am using a customized unlocking system inside Dracut, a very non-stock setup. At any rate between us we have the hiberation supported and hiberation not supported test cases.

**Comment by lukefromdc** [comment]
Not sure of that BUT if the shutdown button is disabled from being added to the panel (w/o my code due to lack of a connection to a session manager,) it also does not appear in the menus

**Comment by lukefromdc** [comment]
They might be from gsm-logout-dialog.c in mate-session-manager, that's where I got the matching icon names from

**Comment by lukefromdc** [comment]
Just got an error dialog working for unsupported cases, will use it for all cases but cancel if the commands fail

**Comment by lukefromdc** [comment]
There is the `--dry-run` option in systemd but it does not work with` -i `or `--ignore-inhibitors` which are necessary to overrride cases of a display manager logged in as root. Also the  `--dry-run` option has been buggy and there have been versions of systemd that ignored it. That could cause someone getting a bad systemd update to get a machine that suspends (first button added) as soon as the panel shut down button is pressed.

**Comment by lukefromdc** [comment]
Finally found a way to hide unsupported buttons, at least when systemd is present and /sys/power/disk exists and lists the supported shutdown/suspend/reboot/hibernate options. When that does not work the buttons are shown unconditionally and an error dialog pops up if an unsupported button is clicked.

**Comment by raveit65** [comment]
See ubuntu ToDo. https://sleeplessbeastie.eu/2020/06/10/how-to-hibernate-system-using-systemd/
One check is 
```
[root@mother rave]# systemctl status systemd-hibernate
○ systemd-hibernate.service - Hibernate
     Loaded: loaded (/usr/lib/systemd/system/systemd-hibernate.service; static)
    Drop-In: /usr/lib/systemd/system/service.d
             └─10-timeout-abort.conf
     Active: inactive (dead)
       Docs: man:systemd-hibernate.service(8)
```
Do you have this service on a system w/o hibernate?
The other file is /etc/systemd/sleep.conf.

**Comment by raveit65** [comment]
I found this check for the size of the hibernation image (from https://docs.kernel.org/admin-guide/pm/sleep-states.html#basic-sysfs-interfaces-for-system-suspend-and-hibernation)
```
[root@mother rave]# cat /sys/power/image_size 
26672320512
```
image_size

    This file controls the size of hibernation images.

    It can be written a string representing a non-negative integer that will be used as a best-effort upper limit of the image size, in bytes. The hibernation core will do its best to ensure that the image size will not exceed that number, but if that turns out to be impossible to achieve, a hibernation image will still be created and its size will be as small as possible. In particular, writing '0' to this file causes the size of hibernation images to be minimum.

    Reading from it returns the current image size limit, which is set to around 2/5 of the available RAM size by default.

Maybe this can be used?

File: mate-panel/panel-action-button.c [hunk 1]
```diff
@@ -213,13 +217,211 @@ panel_action_logout (GtkWidget *widget)
...
+			/*We use the inbuilt gtk response types for simplicity*/
+			suspend_btn = gtk_dialog_add_button (GTK_DIALOG(dialog),
+							     "Suspend",
+							     GTK_RESPONSE_APPLY);
+
+			image = gtk_image_new_from_icon_name("battery", GTK_ICON_SIZE_MENU);
+			gtk_button_set_image (GTK_BUTTON (suspend_btn), image);
+		}
+
+		supported = system ("cat /sys/power/disk | grep hibernate  > /dev/null");
```
Comment (raveit65) [inline mate-panel/panel-action-button.c:362]: I think this file controls what the system should do after the hibernation image is writen.
```
[root@mother rave]# cat /sys/power/disk
[platform] shutdown reboot suspend test_resume 
```
see kernel docs https://docs.kernel.org/admin-guide/pm/sleep-states.html#basic-sysfs-interfaces-for-system-suspend-and-hibernation
My notebook w/o hibernation shows me this
```
[rave@satellite ~]$ cat /sys/power/disk 
[disabled]
```
So you should simply grep for suspend?
Or ifnot [disabled]  ?

**Comment by raveit65** [comment]
I do  not know nothing about power states, same like you.
I only read https://docs.kernel.org/admin-guide/pm/sleep-states.html#basic-sysfs-interfaces-for-system-suspend-and-hibernation
Please read kernel docs
The file /sys/power/disk is the right place.
On a  system w hibernate it looks like this
```
[root@mother rave]# cat /sys/power/disk
[platform] shutdown reboot suspend test_resume 
```
On a system w/o hibernate like yours it look like this
```
[rave@satellite ~]$ cat /sys/power/disk 
[disabled]
```
First output shows only what happens after the hibernate image was writen.
So your checks are wrong.
I am sure you understand kernel docs better than my pidgin english ;)

**Comment by lukefromdc** [comment]
Thanks, I will redo those

**Comment by lukefromdc** [comment]
Also seeing /sys/power/disk controls hibernation ONLY, others are (as seems rational) checked elsewhere

**Comment by raveit65** [comment]
> Also seeing /sys/power/disk controls hibernation ONLY, others are (as seems rational) checked elsewhere

yes, that`s the point.
```
grep disabled /sys/power/disk 
[disabled]
```
Than we know the system doesn't support hibernation.

**Comment by lukefromdc** [comment]
The output I get from cat` /sys/power/disk` is

`[platform] shutdown reboot suspend test_resume`
on a machine not supporting hibernation

**Comment by raveit65** [comment]
Looks like your machine should support hibernation.
I have other results on fedora 38
with hibernation
`[platform] shutdown reboot suspend test_resume`

w/o hibernation
`[disabled]`

Any way, i don't want to look why we have different results from `/sys/power/disk`
For me  an error message is OK when hibernation isn't suported

**Comment by lukefromdc** [comment]
OK, I will show the buttons unconditionally. I am not going to try to set up hibernation on a machine with 48G ram and custom code in dracut. This will also simplify the code

**Comment by raveit65** [comment]
The main reason for hibernation is a security one.
With hibernation support i can unplug the whole current circuit (system, routers, monitors, etc) from power grid when i am not at home ;)
And i save power.

**Comment by lukefromdc** [comment]
I rarely suspend-and never if I am going to be out of the house, always a full shutdown for that. Real point is we need to support many use cases, and our different configurations allow testing two right here

**Comment by lukefromdc** [comment]
Latest force-push shows all buttons unconditionally. Error dialogs for suspend and hiberation failures are now more informative, and they are removed from the shutdown and reboot cases as these should always work unless logout is disabled in gsettings. In that case, the shutdown button cannot be added to the panel and is unresponsive if already there, same behavior as in x11.

**Comment by lukefromdc** [comment]
About the other panel action buttons:
connect to server works without any changes at all
search for files and run application also work w no changes
lock screen and logout depend on session management and do not work as of now

 in wayland screen locking and logout are usually controlled by separate programs. To support these would require an optional build time dependency on a wayland screenlocker and logout manager, or possibly importing code from them to build MATE wayland backends. Simplest way to use such programs though would be a panel launcher for them put in the panel by a user. That would allow user to install any logout manager and screenlocker they want, which may vary with the choice of compositor.

The compositor does a lot of the work in wayland. In my wayfire sessions, wayfire is doing the work of compiz/marco, mate-session-manager, and possibly even some portions of mate-settings-daemon. I have created a local test session that runs from SDDM using just two text files and the edited wayfire.ini file.  There might be other compositors more easily integrated into a session now or in the future plus that in no way would harm the ability to use MATE components in any wlroots-based wayland session. That last part is the same modularity we now have in Xorg where the panel, caja, etc can be used outside of MATE

**Comment by raveit65** [comment]
This works fine. The hibernate button is there and hibernation process works as expected.
But the suspend button is enabled as default. Can this be change to shutdown, same as in x11?

**Comment by raveit65** [comment]
Can you please add the  .linked style class to GtkButtonBox, which is the parent of the buttons?
Menta and Submarine themes use that style class to theme the logout dialog.
Bildschirmfoto zu 2023-09-28 20-31-15

**Comment by lukefromdc** [comment]
Done and done: GtkInspector shows the .linked style class, and the subtle squaring of the inside corners in those themes now works. Also used `gtk_widget_grab_focus` to make the shutdown button the default. Click the panel button, hit return, and the machine shuts down

**Comment by lukefromdc** [comment]
Note that when I used GtkInspector to change the theme of the dialog (and the whole panel as it's part of the panel), the buttons did NOT become connected, just had square inside corners with the .linked style class added to buttonbox. Do I need to add it to the buttons too?

**Comment by raveit65** [comment]
No, `.linked button` is the right syntax in themes. But i see the gap between the buttons too. Also the corner-radius from
left and right buttons will be ignored. It looks better than before. I need to think about..... This can be done later, ihmo.
waylang_shutdown_dialog
What we really need is to center the buttonbox. In Gtk-inspector i choose halign=GTK_ALIGN_CENTER

**Comment by lukefromdc** [comment]
In fact, I get identical style classes between x11 and wayland with this

**Comment by lukefromdc** [comment]
I actually GOT the corner radius, this is what setting the theme for the dialog in GtkInspector to green-submarine looked like. This is a photo as I haven't set up a wayland screenshooter yet.

wayland_shut_down_dialog


Adding the .text-button style class to the buttons had no effect. I am going to have to find the code in the themes and have another look at this.

**Comment by lukefromdc** [comment]
Problem is in the theme and a rendering difference between x11 and wayland: you have
```
* {
    padding: 0;
    background-clip: padding-box;
    -GtkDialog-button-spacing: 4; /* ie. logout dialog */
    -GtkDialog-action-area-border: 10; /* ie. logout dialog */
```
in green-submarine, changing the button spacing to zero aligns the buttons.  I am guessing in x11 the .linked style class overrides that and in wayland it doesn't-or that there is some special way to create linked buttons.

Centering works  fine applying halign=GTK_ALIGN_CENTER to the buttonbox BTW.

**Comment by lukefromdc** [comment]
Refactoring a bit to split out creating the buttons so this is easier to work on

**Comment by lukefromdc** [comment]
This is what I've got the dialog looking like in wayfire with green-submarine. Again, we have
` -GtkDialog-action-area-border: 10; /* ie. logout dialog */`
as a default in green-submarine and probably most of the other themes. It is somehow overridden for linked buttons in x11 but now wayland. Since it is not the default behavior for linked buttons, the .linked style class should set a spacing of zero in themes. I tested that, got a perfect rendering of linked buttons, and centered as well from setting `gtk_widget_set_halign(buttonbox,GTK_ALIGN_CENTER);`

wayland_shutdown_dialog2

**Comment by lukefromdc** [comment]
also tested a build with `--enable-wayland=no` to make sure I didn't mess up the selectors for that, build went fine and ran on x11 as expected. Does not render on wayland w/o wayland support of course. Then rebuilt with wayland support, no surprises.

**Comment by raveit65** [comment]
> Problem is in the theme and a rendering difference between x11 and wayland: you have
> 
> ```
> * {
>     padding: 0;
>     background-clip: padding-box;
>     -GtkDialog-button-spacing: 4; /* ie. logout dialog */
>     -GtkDialog-action-area-border: 10; /* ie. logout dialog */
> ```
> 
> in green-submarine, changing the button spacing to zero aligns the buttons. I am guessing in x11 the .linked style class overrides that and in wayland it doesn't-or that there is some special way to create linked buttons.
>
> Centering works fine applying halign=GTK_ALIGN_CENTER to the buttonbox BTW.

Good analyses, i am sure i can find out in submarine and menta themes why the `linked` style class doesn't override the button-spacing. Dialog looks fine now.

File: mate-panel/panel-action-button.c [hunk 2]
```diff
@@ -213,13 +218,202 @@ panel_action_logout (GtkWidget *widget)
...
+show_error_dialog(GtkWidget *label)
 {
-	PanelSessionManager *manager;
+	GtkWidget *dialog, *box, *content_area, *button, *image;
+	dialog = gtk_dialog_new();
+	button = gtk_dialog_add_button (GTK_DIALOG(dialog),
+			       "Cancel",
+			       GTK_RESPONSE_CANCEL);
+
+	image = gtk_image_new_from_icon_name("process-stop", GTK_ICON_SIZE_MENU);
```
Comment (raveit65) [inline mate-panel/panel-action-button.c:233]: A space is missing here and some other places.
```suggestion
	image = gtk_image_new_from_icon_name ("process-stop", GTK_ICON_SIZE_MENU);
```

File: mate-panel/panel-action-button.c [hunk 3]
```diff
@@ -213,13 +218,202 @@ panel_action_logout (GtkWidget *widget)
...
-	panel_session_manager_request_shutdown (manager);
+		if (response_id == GTK_RESPONSE_CANCEL)
+		{
+			gtk_widget_destroy(dialog);
+			return;
+		}
+
+		if (response_id == GTK_RESPONSE_OK)
+		{
+			gtk_widget_destroy(dialog);
```
Comment (raveit65) [inline mate-panel/panel-action-button.c:259]: same here
```suggestion
			gtk_widget_destroy (dialog);
```

File: mate-panel/panel-action-button.c [hunk 4]
```diff
@@ -213,13 +218,202 @@ panel_action_logout (GtkWidget *widget)
...
+			 *FIXME: we also need a logind equivalent for non-systemd users
+			 *with session managers running as root
+			 */
+			system ("systemctl reboot -i");
+			return;
+		}
+
+		if (response_id == GTK_RESPONSE_REJECT)
+		{
+			gtk_widget_destroy(dialog);
```
Comment (raveit65) [inline mate-panel/panel-action-button.c:286]: same
```suggestion
			gtk_widget_destroy (dialog);
```

File: mate-panel/panel-action-button.c [hunk 5]
```diff
@@ -213,13 +218,202 @@ panel_action_logout (GtkWidget *widget)
...
+						      "Swapfile or partition must be large enough to support hibernation" "\n"
+						      "System and hardware must support hibernation");
+				show_error_dialog (label);
+			}
+			return;
+		}
+
+		if (response_id == GTK_RESPONSE_APPLY)
+		{
+			gtk_widget_destroy(dialog);
```
Comment (raveit65) [inline mate-panel/panel-action-button.c:305]: same
```suggestion
			gtk_widget_destroy (dialog);
```

File: mate-panel/panel-action-button.c [hunk 6]
```diff
@@ -213,13 +218,202 @@ panel_action_logout (GtkWidget *widget)
...
+			/*FIXME: we also need a logind equivalent for non-systemd users
+			 *with session managers running as root
+			 */
+			ret = system ("systemctl suspend -i");
+			if (ret != 0)
+			{
+				GtkWidget *label;
+				label = gtk_label_new("Suspend not supported on this system" "\n" "\n"
```
Comment (raveit65) [inline mate-panel/panel-action-button.c:313]: same
```suggestion
				label = gtk_label_new ("Suspend not supported on this system" "\n" "\n"
```

File: mate-panel/panel-action-button.c [hunk 7]
```diff
@@ -213,13 +218,202 @@ panel_action_logout (GtkWidget *widget)
...
+panel_action_shutdown (GtkWidget *widget)
+{
+#ifdef HAVE_WAYLAND
+	GdkDisplay *display = gdk_screen_get_display (gdk_screen_get_default());
+	if (GDK_IS_WAYLAND_DISPLAY (display))
+	{
+		GtkWidget *dialog, *hbox, *buttonbox, *shutdown_btn, *label;
+		GtkStyleContext *context;
+
+		dialog = gtk_dialog_new_with_buttons("System Shutdown",
```
Comment (raveit65) [inline mate-panel/panel-action-button.c:352]: same
```suggestion
		dialog = gtk_dialog_new_with_buttons ("System Shutdown",
```

File: mate-panel/panel-action-button.c [hunk 8]
```diff
@@ -213,13 +218,202 @@ panel_action_logout (GtkWidget *widget)
...
+		hbox = gtk_box_new (GTK_ORIENTATION_HORIZONTAL, 6);
+
+		label = gtk_label_new ("Shut this system down now?");
+		gtk_label_set_line_wrap (GTK_LABEL (label), TRUE);
+		gtk_label_set_use_markup (GTK_LABEL (label), TRUE);
+		gtk_box_pack_start (GTK_BOX (hbox), label, TRUE, TRUE, 6);
+		gtk_container_set_border_width (GTK_CONTAINER (hbox), 16);
+		gtk_container_add (GTK_CONTAINER (gtk_dialog_get_content_area (GTK_DIALOG (dialog))),
+						  hbox);
+		gtk_dialog_set_default_response(GTK_DIALOG (dialog), GTK_RESPONSE_OK);
```
Comment (raveit65) [inline mate-panel/panel-action-button.c:401]: same
```suggestion
		gtk_dialog_set_default_response (GTK_DIALOG (dialog), GTK_RESPONSE_OK);
```

File: mate-panel/panel-action-button.c [hunk 9]
```diff
@@ -213,13 +218,202 @@ panel_action_logout (GtkWidget *widget)
...
+
+		buttonbox = gtk_widget_get_parent(shutdown_btn);
+		gtk_widget_set_halign(buttonbox,GTK_ALIGN_CENTER);
```
Comment (raveit65) [inline mate-panel/panel-action-button.c:404]: same
```suggestion
		gtk_widget_set_halign (buttonbox,GTK_ALIGN_CENTER);
```

**Comment by raveit65** [comment]
We have some new build warnings.
https://app.travis-ci.com/github/mate-desktop/mate-panel/jobs/610742722#L5291
```
 CC       mate_panel-panel-action-button.o

panel-action-button.c:251:19: warning: comparison of integers of different signs: 'guint' (aka 'unsigned int') and 'int' [-Wsign-compare]

                if (response_id == GTK_RESPONSE_CANCEL)

                    ~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~

panel-action-button.c:257:19: warning: comparison of integers of different signs: 'guint' (aka 'unsigned int') and 'int' [-Wsign-compare]

                if (response_id == GTK_RESPONSE_OK)

                    ~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~

panel-action-button.c:272:19: warning: comparison of integers of different signs: 'guint' (aka 'unsigned int') and 'int' [-Wsign-compare]

                if (response_id == GTK_RESPONSE_ACCEPT)

                    ~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~

panel-action-button.c:284:19: warning: comparison of integers of different signs: 'guint' (aka 'unsigned int') and 'int' [-Wsign-compare]

                if (response_id == GTK_RESPONSE_REJECT)

                    ~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~

panel-action-button.c:303:19: warning: comparison of integers of different signs: 'guint' (aka 'unsigned int') and 'int' [-Wsign-compare]

                if (response_id == GTK_RESPONSE_APPLY)

                    ~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~

5 warnings generated.

panel-action-button.c:275:4: warning: Value stored to 'ret' is never read [deadcode.DeadStores]

                        ret = system ("reboot now");

                        ^     ~~~~~~~~~~~~~~~~~~~~~

1 warning generated.
```

File: mate-panel/panel-action-button.c (hunk 2 above)
Comment (lukefromdc) [reply mate-panel/panel-action-button.c]: Found and fixed in the latest force-push

File: mate-panel/panel-action-button.c (hunk 3 above)
Comment (lukefromdc) [reply mate-panel/panel-action-button.c]: Found and fixed in the latest force-push

File: mate-panel/panel-action-button.c [hunk 10]
```diff
@@ -213,13 +218,202 @@ panel_action_logout (GtkWidget *widget)
...
+			 *FIXME: we also need a logind equivalent for non-systemd users
+			 *with session managers running as root
+			 */
+			system ("systemctl poweroff -i");
+			return;
+		}
+
+		if (response_id == GTK_RESPONSE_ACCEPT)
+		{
+			gtk_widget_destroy(dialog);
```
[the remark this answers is not in the document]
Comment (lukefromdc) [reply mate-panel/panel-action-button.c]: Found and fixed in the latest force-push

File: mate-panel/panel-action-button.c (hunk 4 above)
Comment (lukefromdc) [reply mate-panel/panel-action-button.c]: found and fixed

File: mate-panel/panel-action-button.c [hunk 11]
```diff
@@ -213,13 +218,202 @@ panel_action_logout (GtkWidget *widget)
...
+			/*FIXME: we also need a logind equivalent for non-systemd users
+			 *with session managers running as root
+			 */
+			ret = system ("systemctl hibernate -i");
+			if (ret != 0)
+			{
+				GtkWidget *label;
+				label = gtk_label_new("Hibernation not supported on this system" "\n" "\n"
```
[the remark this answers is not in the document]
Comment (lukefromdc) [reply mate-panel/panel-action-button.c]: Found and fixed

File: mate-panel/panel-action-button.c (hunk 5 above)
Comment (lukefromdc) [reply mate-panel/panel-action-button.c]: Found and fixed

File: mate-panel/panel-action-button.c (hunk 6 above)
Comment (lukefromdc) [reply mate-panel/panel-action-button.c]: Found and fixed

File: mate-panel/panel-action-button.c (hunk 7 above)
Comment (lukefromdc) [reply mate-panel/panel-action-button.c]: Found and fixed

File: mate-panel/panel-action-button.c (hunk 8 above)
Comment (lukefromdc) [reply mate-panel/panel-action-button.c]: found and fixed

File: mate-panel/panel-action-button.c (hunk 9 above)
Comment (lukefromdc) [reply mate-panel/panel-action-button.c]: found and fixed

**Comment by lukefromdc** [comment]
Found a fast way to find ALL instances of no space between a letter and a ( bracket: pluma search for each of the 26 letters followed by an opening bracket with no space. Fixed the ones that were mine, left the ones in code I did not change

**Comment by lukefromdc** [comment]
Still getting the failed distro install from travis:
```
install:

  - sudo apt-get install -y python3-pip python3-setuptools

  - sudo pip3 install --upgrade pip

  - sudo pip install PyGithub

  - ./docker-build --name ${DISTRO} --config .travis.yml --install

The command "./docker-build --name ${DISTRO} --config .build.yml --install" failed and exited with 1 during .

Your build has been stopped.
```

**Comment by raveit65** [comment]
sigh
Those commands are in our docker-build script, see https://github.com/mate-desktop/mate-dev-scripts/blob/master/travis/docker-build#L47
Maybe there is a connection problem to mate-dev repo?
@yetist 
Any idea what could be wrong here?
...from one day to next day

**Comment by raveit65** [comment]
hmm, pygithub and deps were updated.
last working build https://app.travis-ci.com/github/mate-desktop/mate-panel/jobs/610742722#L182
```
pip3 install PyGithub

Collecting PyGithub

  Downloading PyGithub-1.59.1-py3-none-any.whl (342 kB)

Collecting pynacl>=1.4.0

  Downloading PyNaCl-1.5.0-cp36-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl (856 kB)

Requirement already satisfied: requests>=2.14.0 in /usr/lib/python3/dist-packages (from PyGithub) (2.22.0)

Collecting pyjwt[crypto]>=2.4.0

  Downloading PyJWT-2.8.0-py3-none-any.whl (22 kB)

Collecting deprecated

  Downloading Deprecated-1.2.14-py2.py3-none-any.whl (9.6 kB)

Collecting cffi>=1.4.1

  Downloading cffi-1.16.0-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (444 kB)

Collecting cryptography>=3.4.0; extra == "crypto"

  Downloading cryptography-41.0.4-cp37-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.4 MB)

Collecting wrapt<2,>=1.10

  Downloading wrapt-1.15.0-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (81 kB)

Collecting pycparser

  Downloading pycparser-2.21-py2.py3-none-any.whl (118 kB)

Installing collected packages: pycparser, cffi, pynacl, cryptography, pyjwt, wrapt, deprecated, PyGithub

Successfully installed PyGithub-1.59.1 cffi-1.16.0 cryptography-41.0.4 deprecated-1.2.14 pycparser-2.21 pyjwt-2.8.0 pynacl-1.5.0 wrapt-1.15.0
```
build from today https://app.travis-ci.com/github/mate-desktop/mate-panel/jobs/610790065#L182
```
pip3 install PyGithub

Collecting PyGithub

  Downloading PyGithub-2.1.0.post0-py3-none-any.whl (302 kB)

Collecting typing-extensions>=4.0.0

  Downloading typing_extensions-4.8.0-py3-none-any.whl (31 kB)

Collecting Deprecated

  Downloading Deprecated-1.2.14-py2.py3-none-any.whl (9.6 kB)

Collecting python-dateutil

  Downloading python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)

Requirement already satisfied: requests>=2.14.0 in /usr/lib/python3/dist-packages (from PyGithub) (2.22.0)

Requirement already satisfied: urllib3 in /usr/lib/python3/dist-packages (from PyGithub) (1.25.8)

Collecting pynacl>=1.4.0

  Downloading PyNaCl-1.5.0-cp36-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl (856 kB)

Collecting pyjwt[crypto]>=2.4.0

  Downloading PyJWT-2.8.0-py3-none-any.whl (22 kB)

Collecting wrapt<2,>=1.10

  Downloading wrapt-1.15.0-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (81 kB)

Requirement already satisfied: six>=1.5 in /usr/lib/python3/dist-packages (from python-dateutil->PyGithub) (1.14.0)

Collecting cffi>=1.4.1

  Downloading cffi-1.16.0-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (444 kB)

Collecting cryptography>=3.4.0; extra == "crypto"

  Downloading cryptography-41.0.4-cp37-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.4 MB)

Collecting pycparser

  Downloading pycparser-2.21-py2.py3-none-any.whl (118 kB)

Installing collected packages: typing-extensions, wrapt, Deprecated, python-dateutil, pycparser, cffi, pynacl, cryptography, pyjwt, PyGithub

Successfully installed Deprecated-1.2.14 PyGithub-2.1.0.post0 cffi-1.16.0 cryptography-41.0.4 pycparser-2.21 pyjwt-2.8.0 pynacl-1.5.0 python-dateutil-2.8.2 typing-extensions-4.8.0 wrapt-1.15.0
```
I am guessing our docker-build script needs to be updated :/
@yetist
Can you please take a look at it?

**Review by raveit65 — APPROVED** [review]
LGTM,
travis is running again, warnings and formatting is fixed.

**Comment by lukefromdc** [comment]
I now have a working, installable mate-wayland session, for now in my own repo at
https://github.com/lukefromdc/mate-wayland-session
can import it to mate-desktop if desired or work on it there

**Comment by raveit65** [comment]
> I now have a working, installable mate-wayland session, for now in my own repo at https://github.com/lukefromdc/mate-wayland-session can import it to mate-desktop if desired or work on it there

Sure, might be a good first workaround.
In general i think we a new new autostart phase, DisplayServer, that runs before Initialization in mate-session.
something like gnome did years ago.
https://gitlab.gnome.org/GNOME/gnome-session/-/commit/672cb149b846597e7d3cb66692d31a7df5450ee2

**Comment by raveit65** [comment]
Looks like nobody else like to review.
Anyway, let's merge it. I doesn't affect x11.

**Pull request merged**